File: core-service/src/main/java/com/yoku/coreservice/Service/UserService.java
package com.yoku.coreservice.Service;

import com.yoku.coreservice.Entity.User.UserProfile;
import com.yoku.coreservice.Exceptions.InvalidArgumentException;
import com.yoku.coreservice.Exceptions.UserNotFoundException;
import com.yoku.coreservice.Repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.UUID;

@Service
public class UserService {

    @Autowired
    private UtilService utilService;

    @Autowired
    private UserRepository userRepository;

    private UserProfile FindUserProfileOrThrow(String userId) throws UserNotFoundException {
        return userRepository.findByUserId(UUID.fromString(userId))
                .orElseThrow(() -> new UserNotFoundException("User Profile not found with User Id:" + userId));
    }

    private UserProfile FindUserProfileOrThrow(UUID userId) throws UserNotFoundException {
        return userRepository.findByUserId(userId)
                .orElseThrow(() -> new UserNotFoundException("User Profile not found with User Id:" + userId));
    }

    public UserProfile GetUserProfile(String userId) throws InvalidArgumentException, UserNotFoundException {
        // Validate UUID
        utilService.ValidateUUIDFromString(userId);

        // Find User Profile from Database else throw if null
        return FindUserProfileOrThrow(userId);

    }

    public UserProfile UpdateUserProfile(UserProfile user) throws InvalidArgumentException, UserNotFoundException{
        return userRepository.save(user);
    }



}

File: src/main/kotlin/com/yoku/colovia/service/user/UserService.kt
package com.yoku.colovia.service.user

import com.yoku.colovia.entity.dto.UserDTO
import com.yoku.colovia.entity.dto.UserPartialDTO
import com.yoku.colovia.entity.user.UserProfile
import com.yoku.colovia.exceptions.UserNotFoundException
import com.yoku.colovia.repository.UserRepository
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.*

@Service
class UserService(
    private val userRepository: UserRepository,
    private val cachedUserService: CachedUserService
) {
    private val logger: Logger = LoggerFactory.getLogger(UserService::class.java)

    @Throws(UserNotFoundException::class)
    fun getUserProfileByEmail(email: String): UserProfile {
        return userRepository.findByEmail(email)
            .orElseThrow{UserNotFoundException("User not found with email: $email")}
    }

    @Throws(UserNotFoundException::class)
    fun getUserProfileById(userId: UUID): UserProfile {
        return cachedUserService.getUserProfileById(userId)
    }

    fun getBatchUserProfilesByIds(userIds: List<UUID>): Map<UUID, UserPartialDTO?> {
        return userIds.associateWith { userId ->
            try {
                cachedUserService.getUserProfileById(userId).toPartialDTO()
            } catch (e: UserNotFoundException) {
                logger.error("User not found with ID: $userId")
                null
            }
        }
    }

    @Throws(UserNotFoundException::class)
    fun updateUserProfile(user: UserDTO): UserProfile {
        return cachedUserService.updateUserProfile(user)
    }
}
